### PR TITLE
api, chunk: progress bar support

### PR DESCRIPTION
**Pull request by jmozah**
This PR provides the HTTP api support to retrieve tag counters.

A new endpoint /bzz-tag/<swarm hash> is added to achieve this.
Instead of supplying the swarm hash, the tagID can also be supplied like /bzz-tag/?tagID=<tagid>
The tag id of a uploaded file is returned in the response header.
Any client which wants to implement a progress bar will get this tagId and pull the counters using this API.

### CODE REVIEW

File: api/http/server.go [hunk 1]
```diff
@@ -865,6 +875,57 @@ func (s *Server) HandleGetFile(w http.ResponseWriter, r *http.Request) {
...
 }
 
+// HandleGetTag responds to the following request
+//    - bzz-tag:/<manifest>  and
+//    - bzz-tag:/?tagId=<tagId>
+// Clients should use root hash or the tagID to get the tag counters
+func (s *Server) HandleGetTag(w http.ResponseWriter, r *http.Request) {
+	getTagCount.Inc(1)
+	uri := GetURI(r.Context())
+	fileAddr := uri.Address()
```
Comment (acud) [inline api/http/server.go:944]: `uri` could be nil, thus you are risking a panic here. you won't see it though, since the http server catches panics (but logs them)

File: api/http/server.go [hunk 2]
```diff
@@ -865,6 +875,57 @@ func (s *Server) HandleGetFile(w http.ResponseWriter, r *http.Request) {
...
+
+	var tag *chunk.Tag
+	if fileAddr == nil {
+		tagString := r.URL.Query().Get("tagId")
+		if tagString == "" {
+			getTagFail.Inc(1)
+			respondError(w, r, "Missing one of the mandatory argument", http.StatusBadRequest)
```
Comment (acud) [inline api/http/server.go:951]: return after this statement

File: api/http/server.go [hunk 3]
```diff
@@ -865,6 +875,57 @@ func (s *Server) HandleGetFile(w http.ResponseWriter, r *http.Request) {
...
+		}
+
+		u64, err := strconv.ParseUint(tagString, 10, 32)
+		if err != nil {
+			getTagFail.Inc(1)
+			respondError(w, r, "Invalid tagId argument", http.StatusBadRequest)
```
Comment (acud) [inline api/http/server.go]: return after this

File: api/http/server.go [hunk 4]
```diff
@@ -865,6 +875,57 @@ func (s *Server) HandleGetFile(w http.ResponseWriter, r *http.Request) {
...
+		}
+		tagId := uint32(u64)
+
+		tag, err = s.api.Tags.Get(tagId)
+		if err != nil {
+			getTagNotFound.Inc(1)
+			respondError(w, r, "Tag not found", http.StatusNotFound)
```
Comment (acud) [inline api/http/server.go:966]: return

File: api/http/server.go [hunk 5]
```diff
@@ -865,6 +875,57 @@ func (s *Server) HandleGetFile(w http.ResponseWriter, r *http.Request) {
...
+		}
+	} else {
+
+		tagByFile, err := s.api.Tags.GetByAddress(fileAddr)
+		if err != nil {
+			getTagNotFound.Inc(1)
+			respondError(w, r, "Tag not found", http.StatusNotFound)
```
Comment (acud) [inline api/http/server.go:974]: return

File: api/http/server.go [hunk 6]
```diff
@@ -865,6 +875,57 @@ func (s *Server) HandleGetFile(w http.ResponseWriter, r *http.Request) {
...
+		}
+		tag = tagByFile
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Header().Set("Cache-Control", "no-cache, private, max-age=0")
+	r.Header.Del("ETag")
+	w.WriteHeader(http.StatusOK)
+	bytesToSend, err := tag.MarshalBinary()
```
Comment (acud) [inline api/http/server.go]: `MarshalBinary` gives back a byte slice.
use `json.Marshal()` instead

File: state/dbstore.go [hunk 7]
```diff
@@ -34,6 +36,7 @@ type Store interface {
 	Get(key string, i interface{}) (err error)
 	Put(key string, i interface{}) (err error)
 	Delete(key string) (err error)
+	Iterate(prefix string, iterFunc iterFunction) (err error)
```
Comment (acud) [inline state/dbstore.go:38]: this should already be on master. please rebase

**Comment by acud** [comment]
@jmozah i added some comments in the body of the PR.
also - please rebase on `master`

File: api/http/server.go (hunk 6 above)
Comment (jmozah) [reply api/http/server.go]: This will need that i export all the fields of Tag. That means i change many files. Is it okay?

File: state/dbstore.go (hunk 7 above)
Comment (jmozah) [reply state/dbstore.go:38]: Rebased.

File: api/http/server.go (hunk 5 above)
Comment (jmozah) [reply api/http/server.go:974]: Done

File: api/http/server.go (hunk 4 above)
Comment (jmozah) [reply api/http/server.go:966]: Done

File: api/http/server.go (hunk 3 above)
Comment (jmozah) [reply api/http/server.go]: Done

File: api/http/server.go (hunk 2 above)
Comment (jmozah) [reply api/http/server.go:951]: oops.. Done

File: api/http/server.go (hunk 1 above)
Comment (jmozah) [reply api/http/server.go:944]: I have added it. 

Note: This is like this in all other functions.. May be we should add that everywhere too.

File: api/http/server.go (hunk 6 above)
Comment (acud) [reply api/http/server.go]: not necessary. just have `Tag` implement `MarshalJSON`/`UnmarshalJSON` ([link](https://play.golang.org/p/kN-_1lLoggl))

Comment (janos) [reply api/http/server.go]: Also, MarshalBinary returns the binary representation fo tag, but the json content type header is sent. I think that sending json would be more appropriate.

Comment (jmozah) [reply api/http/server.go]: Now the API returns stuff in json encoded format.

**Comment by jmozah** [comment]
@zelig In the GET API.. i currently send the entire tag structure as it is.  @acud was talking about sending counters only based on Status(). I don't get it. Why should client be bothered about states? From the client perspective, he gets the counters and he decides what to do with them. 

This is the only thing remaining in this PR for now.

File: chunk/tag.go [hunk 8]
```diff
@@ -216,3 +218,57 @@ func decodeInt64Splice(buffer *[]byte) int64 {
 	*buffer = (*buffer)[n:]
 	return val
 }
+
+// marshall friendly tag structure
+type MarshallTag struct {
```
Comment (janos) [inline chunk/tag.go]: I think that this type can be unexported.

Comment (jmozah) [reply chunk/tag.go]: Done

Comment (janos) [reply chunk/tag.go]: Thanks.

File: api/http/server.go [hunk 9]
```diff
@@ -897,6 +907,66 @@ func (s *Server) HandleGetFile(w http.ResponseWriter, r *http.Request) {
...
+	if uri == nil {
+		getTagFail.Inc(1)
+		respondError(w, r, "Error decoding uri", http.StatusBadRequest)
+		return
+	}
+	fileAddr := uri.Address()
+
+	var tag *chunk.Tag
+	if fileAddr == nil {
+		tagString := r.URL.Query().Get("tagId")
```
Comment (janos) [inline api/http/server.go]: Maybe to rename `tagId` to `id`? This is already on bzz-tag scheme, so tag is implicit.

**Review by acud — DISMISSED** [review]
LGTM

File: api/http/server.go (hunk 9 above)
Comment (jmozah) [reply api/http/server.go]: Done

File: chunk/tags.go [hunk 10]
```diff
@@ -76,6 +77,28 @@ func (ts *Tags) Get(uid uint32) (*Tag, error) {
 	return t.(*Tag), nil
 }
 
+// GetByAddress returns the underlying tag for the address or an error if not found
+func (ts *Tags) GetByAddress(address []byte) (*Tag, error) {
+	t := &Tag{}
```
Comment (acud) [inline chunk/tags.go]: can probably be simplified to:
```go
var t *Tag
ts.tags.Range(func(key interface{}, value interface{}) bool {
	rcvdTag := value.(*Tag)
	if bytes.Equal(rcvdTag.Address, address) {
		t = rcvdTag
		return false
	}
	return true
})
if t == nil {
        return nil,errTagNotFound
}
return t,nil
```

Comment (jmozah) [reply chunk/tags.go]: Done

**Review by acud — APPROVED** [review]
lgtm

File: chunk/tag.go [hunk 11]
```diff
@@ -216,3 +218,57 @@ func decodeInt64Splice(buffer *[]byte) int64 {
 	*buffer = (*buffer)[n:]
 	return val
 }
+
+// marshall friendly tag structure
+type jsonTag struct {
```
Comment (zelig) [inline chunk/tag.go]: let's just make the `Address` field a string field on `Tag` and remove custom json code. 
or define `(Un)MarshalJSON` on `Address` type.

File: chunk/tag_test.go [hunk 12]
```diff
@@ -271,3 +271,47 @@ func TestMarshallingNoAddr(t *testing.T) {
 		t.Fatalf("expected tag addresses to be equal length")
 	}
 }
+
+func TestJsonMarshallingUnMarshalling(t *testing.T) {
```
Comment (zelig) [inline chunk/tag_test.go]: comment...

File: state/dbstore.go [hunk 13]
```diff
@@ -107,14 +107,25 @@ func (s *DBStore) Delete(key string) (err error) {
...
 }
 
-// Iterate entries which has a specific prefix
-// The key and value may be modified in the iterFunc
-func (s *DBStore) Iterate(prefix string, iterFunc func([]byte, []byte)) (err error) {
+// iterFunction is a function called on every key/value pair obtained
+// If true is returned in the stop variable, iteration will
+// stop, and by returning the error, that error will be
+// propagated to the called iterator method on Iterate.
+type iterFunction func([]byte, []byte) (stop bool, err error)
```
Comment (zelig) [inline state/dbstore.go]: i recommend named variables here like `(key, value []byte)`, easier to interpret

**Review by zelig — CHANGES_REQUESTED** [review]
minor comments only
plus, could we get rid of custom marshalling?

File: state/dbstore.go (hunk 13 above)
Comment (zelig) [reply state/dbstore.go]: This functionality is added also in https://github.com/ethersphere/swarm/pull/1554/files#r313813885

File: api/http/server.go [hunk 14]
```diff
@@ -62,11 +62,14 @@ var (
...
 	getListFail     = metrics.NewRegisteredCounter("api.http.get.list.fail", nil)
+	getTagCount     = metrics.NewRegisteredCounter("api.http.get.tag.count", nil)
+	getTagNotFound  = metrics.NewRegisteredCounter("api.http.get.tag.notfound", nil)
+	getTagFail      = metrics.NewRegisteredCounter("api.http.get.tag.fail", nil)
 )
 
 const (
-	SwarmTagHeaderName = "x-swarm-tag" // Presence of this in header indicates the tag
-	PinHeaderName      = "x-swarm-pin" // Presence of this in header indicates pinning required
+	TagHeaderName = "x-swarm-tag"
```
[the remark this answers is not in the document]
Comment (jmozah) [reply api/http/server.go]: This is before the merge from master .. it was commented  in "SwarmTagHeaderName" ...
Now only "TagHeaderName" exists and is commented.

File: chunk/tag.go (hunk 11 above)
Comment (jmozah) [reply chunk/tag.go]: Done

File: chunk/tag_test.go (hunk 12 above)
Comment (jmozah) [reply chunk/tag_test.go]: Done

File: state/dbstore.go (hunk 13 above)
Comment (jmozah) [reply state/dbstore.go]: Done.

This function was pulled from there only.

File: chunk/tag.go (hunk 11 above)
Comment (zelig) [reply chunk/tag.go]: no. I mean on the `Tag`, the exported fields should be exactly these, so you can just JSON marshal it without using any custom marshaller.

**Review by zelig — CHANGES_REQUESTED** [review]
please remove custom marshalling it is unnecessary

File: chunk/tag.go [hunk 15]
```diff
@@ -45,7 +47,7 @@ const (
 type Tag struct {
 	Uid       uint32    // a unique identifier for this tag
 	Name      string    // a name tag for this tag
-	Address   Address   // the associated swarm hash for this tag
+	Address   string    // the associated swarm hash for this tag
```
Comment (acud) [inline chunk/tag.go]: why this change?

File: chunk/tag.go [hunk 16]
```diff
@@ -113,7 +115,7 @@ func (t *Tag) Total() int64 {
 func (t *Tag) DoneSplit(address Address) int64 {
 	total := atomic.LoadInt64(&t.split)
 	atomic.StoreInt64(&t.total, total)
-	t.Address = address
+	t.Address = hex.EncodeToString(address)
```
Comment (acud) [inline chunk/tag.go]: ditto

**Review by acud — CHANGES_REQUESTED** [review]
not sure why the change to `GetByAddress` to handle `string`. should handle `Address` as was before

File: chunk/tag.go (hunk 11 above)
Comment (jmozah) [reply chunk/tag.go]: did you not mean making "Address" in Tag to string?

Comment (acud) [reply chunk/tag.go]: He meant to have the fields on `Tag` exported, then add a custom json marshal/unmarshal on `chunk.Address`, that way the built in json marshaler will serialize/deserialize the exported fields, but will pick up the `Address` type and call the custom marshaler function you define on it to convert from/to string

**Comment by zelig** [comment]
@jmozah see linting error https://travis-ci.org/ethersphere/swarm/jobs/572356780#L491

File: chunk/tag.go (hunk 11 above)
Comment (acud) [reply chunk/tag.go]: @zelig an idea about how to serialize the response:
```js
{
  tag: id
  stored: {
    remaining: 100
    total: 1000
  }
  split: {
    remaining: 100
    total: 1000
  }
  ...
}
```
if we just serialize the struct as is, we end up delegating the logic which is in `Tag.Status` to the client:
```go
// Status returns the value of state and the total count
func (t *Tag) Status(state State) (int64, int64, error) {
	count, seen, total := t.Get(state), atomic.LoadInt64(&t.seen), atomic.LoadInt64(&t.total)
	if total == 0 {
		return count, total, errNA
	}
	switch state {
	case StateSplit, StateStored, StateSeen:
		return count, total, nil
	case StateSent, StateSynced:
		stored := atomic.LoadInt64(&t.stored)
		if stored < total {
			return count, total - seen, errNA
		}
		return count, total - seen, nil
	}
	return count, total, errNA
}
```
i am not sure that the approach of serializing the struct as is is the way to go as of such.

Comment (jmozah) [reply chunk/tag.go]: @zelig ... I did exported the fields of Tags and @acud didn't like that design. He is of the opinion that we should not export fields of Tag.  He wanted to have a separate JSON Marshaller and UnMarsheller.

Comment (zelig) [reply chunk/tag.go]: > @zelig ... I did exported the fields of Tags and @acud didn't like that design. He is of the opinion that we should not export fields of Tag. He wanted to have a separate JSON Marshaller and UnMarsheller.

I all for code simplicity,  the tag  struct contains all the info, we dont need to second guess what the renderer will massage on the data..
@acud if you insist, we  could have  it your way.

Anyhow we should continue  with http upload responses using the  tags

File: chunk/tag.go (hunk 15 above)
Comment (jmozah) [reply chunk/tag.go]: Misunderstood the comment. changing it back to Address.

File: chunk/tag.go (hunk 16 above)
Comment (jmozah) [reply chunk/tag.go]: Misunderstood the comment. changing it back to Address.

File: chunk/tag.go (hunk 11 above)
Comment (jmozah) [reply chunk/tag.go]: ok. I am exporting all the fields of Tags and closing it for now.

**Review by acud — APPROVED** [review]

**Review by janos — APPROVED** [review]

**Review by zelig — APPROVED** [review]
